agent/taskresource/envFiles/envfile.go

```diff
@@ -0,0 +1,198 @@
+
+	// env file related attributes
+	envfileLocations []apicontainer.EnvironmentFile // list of env file urls
+
+	executionCredentialsID string
+	credentialsManager     credentials.Manager
+	s3ClientCreator        factory.S3ClientCreator
+	os                     oswrapper.OS
+	ioutil                 ioutilwrapper.IOUtil
+
+	// Fields for the common functionality of task resource. Access to these fields are protected by lock.
+	createdAtUnsafe     time.Time
+	desiredStatusUnsafe resourcestatus.ResourceStatus
+	knownStatusUnsafe   resourcestatus.ResourceStatus
+	appliedStatusUnsafe resourcestatus.ResourceStatus
+	statusToTransitions map[resourcestatus.ResourceStatus]func() error
+	terminalReason      string
+	terminalReasonOnce  sync.Once
+	lock                sync.RWMutex
+}
+
+func NewEnvironmentFileResource(cluster, taskARN, region, dataDirOnHost string, credentialsManager credentials.Manager,
+	executionCredentialsID string) (*EnvironmentFileResource, error) {
+	envfileResource := &EnvironmentFileResource{
+		cluster:                cluster,
+		taskARN:                taskARN,
+		region:                 region,
+		os:                     oswrapper.NewOS(),
+		ioutil:                 ioutilwrapper.NewIOUtil(),
+		s3ClientCreator:        factory.NewS3ClientCreator(),
+		executionCredentialsID: executionCredentialsID,
+		credentialsManager:     credentialsManager,
+	}
+
+	taskARNFields := strings.Split(taskARN, "/")
+	taskID := taskARNFields[len(taskARNFields)-1]
+	// we save envfiles to path: /var/lib/ecs/data/envfiles/cluster_name/task_id/${s3filename.env}
+	envfileResource.resourceDir = filepath.Join(dataDirOnHost, envFileDirPath, cluster, taskID)
+
+	return envfileResource, nil
+}
+
+func (envfile *EnvironmentFileResource) Initialize(resourceFields *taskresource.ResourceFields,
+	taskKnownStatus status.TaskStatus,
+	taskDesiredStatus status.TaskStatus) {
+	envfile.lock.Lock()
+	defer envfile.lock.Unlock()
+
+	envfile.initStatusToTransition()
+	envfile.credentialsManager = resourceFields.CredentialsManager
+	envfile.s3ClientCreator = factory.NewS3ClientCreator()
+	envfile.os = oswrapper.NewOS()
+	envfile.ioutil = ioutilwrapper.NewIOUtil()
+}
+
+func (envfile *EnvironmentFileResource) initStatusToTransition() {
+	resourceStatusToTransitionFunc := map[resourcestatus.ResourceStatus]func() error{
+		resourcestatus.ResourceStatus(EnvFileCreated): envfile.Create,
+	}
+	envfile.statusToTransitions = resourceStatusToTransitionFunc
+}
+
+func (envfile *EnvironmentFileResource) GetTerminalReason() string {
+	envfile.lock.RLock()
+	defer envfile.lock.RUnlock()
+
+	return envfile.terminalReason
+}
+
+func (envfile *EnvironmentFileResource) setTerminalReason(reason string) {
+	envfile.terminalReasonOnce.Do(func() {
+		seelog.Infof("envfile resource: setting terminal reason for task: [%s]", envfile.taskARN)
+		envfile.terminalReason = reason
+	})
+}
+
+// resource creation, retrieve env file contents in parallel
```

Comment:
Can we add a bit more details in the comment about the env file is downloaded and save on the host in the function?

Reply:
yes my bad, let me add that in

Reply:
actually going to do this in next PR.